runtimes/peregrine/src/lib.rs

```diff
@@ -1456,3 +1461,52 @@ cumulus_pallet_parachain_system::register_validate_block! {
 	BlockExecutor = cumulus_pallet_aura_ext::BlockExecutor::<Runtime, Executive>,
 	CheckInherents = CheckInherents,
 }
+
+struct DidRootDispatcher<Runtime>(sp_std::marker::PhantomData<Runtime>);
+
+impl<Runtime: cumulus_pallet_parachain_system::Config>
+	did::traits::DidRootDispatcher<DidIdentifier, AccountId, Hash, MultiLocation> for DidRootDispatcher<Runtime>
+{
+	fn dispatch(
+		action: did::traits::DidRootStateAction<DidIdentifier, Hash>,
+		dispatcher: AccountId,
+		location: MultiLocation,
+	) -> Result<Weight, sp_runtime::DispatchError> {
+		let asset = MultiAsset {
+			id: Concrete(
+				MultiLocation::here()
+					.pushed_with_interior(AccountId32 {
+						network: Any,
+						id: dispatcher.into(),
+					})
+					.unwrap(),
+			),
+			fun: Fungibility::Fungible(1_000_000_000),
+		};
+		let calls = Xcm(vec![
+			WithdrawAsset(asset.clone().into()),
+			BuyExecution {
+				fees: asset,
+				weight_limit: Unlimited,
+			},
+			Transact {
+				origin_type: OriginKind::Xcm,
+				require_weight_at_most: 1_000_000_000,
+				// TODO: Replace with actual call on target chain
+				call: action.encode().into(),
```

Comment:
this could potentially encode differently on every chain? So we would need to distinguish the target chain?

Reply:
The action is always encoded/decoded in the same way, if the two chains use the same struct definition. We will have to support versioning, of course, which is not yet here. But I also not yet 100& if a transact can contain a generic payload that the other chain can parse and interpret, or if it always has to be an call encoded in the target chain's logic.

Reply:
AFAIK it always has to be a call. But would be good to check this assumption.

Reply:
We will have our own pallet deployed on the target chain, so no problem anyway.

Reply:
But the pallet will not have the same pallet index on every chain. So the encoding of the call will be different.